### PR TITLE
Adding stochastic beam search

### PR DESCRIPTION
**Pull request by jvstokes**
## Description
Adding stochastic beam search to `beam_search.py`

## Changes
`allennlp/nn/samplers/samplers.py`: added `GumbelMaxSampler`
`allennlp/nn/beam_search.py`: added `stochastic_beam_search`

## Tests
`tests/nn/beam_search_tests.py`: added basic tests for stochastic beam search

### CODE REVIEW

File: allennlp/nn/beam_search.py [hunk 1]
```diff
@@ -222,6 +223,11 @@ def _search(
                 start_predicted_classes,
             ) = start_class_log_probabilities.topk(self.beam_size)
 
+        if isinstance(self.sampler, GumbelMaxSampler):
+            last_true_probabilities = torch.gather(
+                start_class_log_probabilities, 1, start_predicted_classes
+            )
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:229]: I actually think we should always gather `last_true_log_probabilities` since there is no guarantee with other samplers that they won't perturb the log probabilities.

```suggestion
        # Some samplers add perturb the log probabilities with random noise, such
        # as the GumbelMaxSampler, so we need to recover and keep track of the
        # true log probabilities here.
        last_true_log_probabilities = torch.gather(
            start_class_log_probabilities, 1, start_predicted_classes
        )
```

Would need a similar change below as well.

File: allennlp/nn/beam_search.py [hunk 2]
```diff
@@ -327,6 +333,26 @@ def _search(
...
+                summed_true_log_probabilities = (
+                    cleaned_log_probabilities.gather(1, predicted_classes)
+                    + expanded_true_log_probabilities
+                )
+
+                # shape: (batch_size, beam_size * per_node_beam_size)
+                reshaped_summed_true_probabilities = summed_true_log_probabilities.reshape(
+                    batch_size, self.beam_size * self.per_node_beam_size
+                )
+
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:355]: Same comment as above.

File: allennlp/nn/beam_search.py [hunk 3]
```diff
@@ -344,6 +370,15 @@ def _search(
...
 
+            # track the true probabilities for the selected beams
+            # shape: (batch_size, beam_size)
+            if isinstance(self.sampler, GumbelMaxSampler):
+                restricted_true_probabilities = reshaped_summed_true_probabilities.gather(
+                    1, restricted_beam_indices
+                )
+
+                last_true_probabilities = restricted_true_probabilities
+
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:381]: Same comment here as well.

File: allennlp/nn/beam_search.py [hunk 4]
```diff
@@ -408,6 +443,10 @@ def _search(
         # shape: (batch_size, beam_size, max_steps)
         all_predictions = torch.cat(list(reversed(reconstructed_predictions)), 2)
 
+        # If stochastic beam search, return the true probabilities instead of perturbed
+        if isinstance(self.sampler, GumbelMaxSampler):
+            return all_predictions, last_true_probabilities
+
         return all_predictions, last_log_probabilities
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:450]: With the above changes, we'd just always return `last_true_log_probabilities`.

```suggestion
        return all_predictions, last_true_log_probabilities
```

File: allennlp/nn/beam_search.py [hunk 5]
```diff
@@ -489,7 +528,35 @@ def top_p_sampling(
...
+        per_node_beam_size: int = None,
+        temperature: float = 1.0,
+    ) -> "BeamSearch":
+        """
+        Given an index of the end token in target vocabulary, return a `BeamSearch` object
+        that can be used to find `beam_size` candidate sequences, found by using Stochastic
+        Beam Search, which leverages Gumbel-TopK sampling. See
+        [*Stochastic Beams and Where to Find Them: The Gumbel-Top-k Trick for Sampling
+        Sequences Without Replacement*, W Kool, H Van Hoof and M Welling, 2010]
+        (https://arxiv.org/abs/1903.06059).
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:546]: I don't think this will render properly in the documentation. It might have to look like:

```suggestion
        [*Stochastic Beams and Where to Find Them: The Gumbel-Top-k Trick for Sampling Sequences Without Replacement*, W Kool, H Van Hoof and M Welling, 2010]
        (https://arxiv.org/abs/1903.06059).
```

You can check this by building and serving the docs locally with `make serve-docs` and then opening [http://127.0.0.1:8000/api/nn/beam_search/](http://127.0.0.1:8000/api/nn/beam_search/) in your browser.

File: allennlp/nn/beam_search.py (hunk 1 above)
Comment (jvstokes) [reply allennlp/nn/beam_search.py:229]: Should I still check `if self.sampler is not None:` ? Right now only the gumbel sampler will return perturbed log probabilities, and deterministic beam search will always keep track of the true probabilities

Comment (epwalsh) [reply allennlp/nn/beam_search.py:229]: Ah true, yes let's do that

File: allennlp/nn/beam_search.py (hunk 5 above)
Comment (jvstokes) [reply allennlp/nn/beam_search.py:546]: This looks fine when viewing the docs, and the linter complains otherwise so I left it!

File: allennlp/nn/beam_search.py [hunk 6]
```diff
@@ -222,6 +223,14 @@ def _search(
                 start_predicted_classes,
             ) = start_class_log_probabilities.topk(self.beam_size)
 
+        # Some samplers add perturb the log probabilities with random noise, such
+        # as the GumbelMaxSampler, so we need to recover and keep track of the
+        # true log probabilities here.
+        if self.sampler is not None:
+            last_true_probabilities = torch.gather(
+                start_class_log_probabilities, 1, start_predicted_classes
+            )
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:232]: I think it's better to define this variable outside of this branch. I'm surprised Python allows this kind of thing, but then again... it's Python so I guess anything goes 😝

```suggestion
        last_true_probabilities: Optional[torch.Tensor] = None
        if self.sampler is not None:
            last_true_probabilities = torch.gather(
                start_class_log_probabilities, 1, start_predicted_classes
            )
```

File: allennlp/nn/beam_search.py [hunk 7]
```diff
@@ -327,6 +336,27 @@ def _search(
                 batch_size, self.beam_size * self.per_node_beam_size
             )
 
+            # Track the true probabilities in addition to perturbed probabilities
+            # for use in samplers that may perturb probabilities
+            if self.sampler is not None:
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:341]: With the above change, might make sense to check against null value of `last_true_probabilities` here to make mypy happy:

```suggestion
            if last_true_probabilities is not None:
```

File: allennlp/nn/beam_search.py [hunk 8]
```diff
@@ -344,6 +374,15 @@ def _search(
                 1, restricted_beam_indices
             )
 
+            # track the true probabilities for the selected beams
+            # shape: (batch_size, beam_size)
+            if self.sampler is not None:
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:379]:
```suggestion
            if last_true_probabilities is not None:
```

File: allennlp/nn/beam_search.py [hunk 9]
```diff
@@ -408,6 +447,10 @@ def _search(
         # shape: (batch_size, beam_size, max_steps)
         all_predictions = torch.cat(list(reversed(reconstructed_predictions)), 2)
 
+        # If a sampler is used, return the true probabilities instead of perturbed probabilities
+        if self.sampler is not None:
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:451]: Same here

```suggestion
        if last_true_probabilities is not None:
```

File: allennlp/nn/beam_search.py (hunk 6 above)
Comment (epwalsh) [reply allennlp/nn/beam_search.py:232]: Also, should this be `last_true_log_probabilities`, not `last_true_probabilities`?

Comment (jvstokes) [reply allennlp/nn/beam_search.py:232]: Yup I'll change that variable name!

File: tests/nn/beam_search_test.py [hunk 10]
```diff
@@ -277,6 +277,26 @@ def test_top_k_search(self):
...
+        beam_size = beam_size or 1
+        batch_size = 5
+
+        # top_p should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(top_k.size())[:-1] == [batch_size, beam_size]
+
+        assert ((0 <= top_k) & (top_k <= 5)).all()
+
+        # log_probs should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(log_probs.size()) == [batch_size, beam_size]
```
Comment (epwalsh) [inline tests/nn/beam_search_test.py:298]: We should be able to verify exactly what `log_probs` should be given `top_k`, which would be good to do in this test. If that's too difficult though, we could have a similar test using a dummy sampler that does the same thing as deterministic beam search. Then we'd just have to check that the `log_probs` result is the same when using the dummy sampler vs using no sampler.

File: allennlp/nn/samplers/samplers.py [hunk 11]
```diff
@@ -123,3 +130,42 @@ def __call__(
...
+@Sampler.register("gumbel-max")
+class GumbelMaxSampler(Sampler):
+    """
+    Represents a `Sampler` which uses the Gumbel-Max trick to sample `num_samples`
+    instances without replacement
+    `logits` is a tensor of log-probabilities to be selected from.
+    `temperature` modules the probabilitis of the selected tokens. A `temperature` below 1.0 produces a
+    sharper probability distribution and a `temperature` above 1.0 produces a flatter probability
+    distribution.
+    `num_samples` is the number of instances to sample
```
Comment (epwalsh) [inline allennlp/nn/samplers/samplers.py:144]: Maybe we should mention here that it returns perturbed log probs?

File: allennlp/nn/samplers/samplers.py [hunk 12]
```diff
@@ -17,22 +17,26 @@ def __init__(self, temperature: float = 1.0, filter_val: float = -float("inf"))
...
     def __call__(
-        self, logits: torch.Tensor, num_samples: int = 1, with_replacement: bool = True
+        self,
+        log_probs: torch.Tensor,
+        perturbed_log_probs: torch.Tensor = None,
+        num_samples: int = 1,
+        with_replacement: bool = True,
     ) -> torch.Tensor:
-        probabilities = torch.nn.functional.softmax(logits, dim=-1)
+        probabilities = torch.nn.functional.softmax(log_probs, dim=-1)
```
Comment (epwalsh) [inline allennlp/nn/samplers/samplers.py:26]: `log_probs` are already normalized, so we can just do `log_probs.exp()` to get `probabilities`.

File: allennlp/nn/beam_search.py [hunk 13]
```diff
@@ -213,15 +214,33 @@ def _search(
...
+                phi = start_class_log_probabilities
+
+                # Our sampler gets the perturbed log probabilities for each token
+                start_top_log_probabilities, start_predicted_classes = self.sampler(
+                    phi, num_samples=self.beam_size
+                )
+            else:
+                start_top_log_probabilities, start_predicted_classes = self.sampler(
+                    start_class_log_probabilities, num_samples=self.beam_size
+                )
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:228]: Is this special case handling for `GumbelMaxSampler` still necessary?

And it looks like the way we call `self.sampler()` in the `else` branch is incorrect / inconsistent with the signature of `Sampler`.

File: allennlp/nn/beam_search.py [hunk 14]
```diff
@@ -489,7 +546,34 @@ def top_p_sampling(
             sampler=sampler_p,
         )
 
+    @classmethod
+    def stochastic_beam_search(
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:550]: Maybe just call this ⬇️ ?

```suggestion
    def stochastic(
```

File: allennlp/nn/beam_search.py [hunk 15]
```diff
@@ -489,7 +546,34 @@ def top_p_sampling(
...
+            beam_size=beam_size,
+            per_node_beam_size=per_node_beam_size or beam_size,
+            sampler=gumbel_sampler,
+        )
+
 
 BeamSearch.register("without_sampling", constructor="without_sampling")(BeamSearch)
 BeamSearch.register("top_p_sampling", constructor="top_p_sampling")(BeamSearch)
 BeamSearch.register("top_k_sampling", constructor="top_k_sampling")(BeamSearch)
+BeamSearch.register("stochastic_beam_search", constructor="stochastic_beam_search")(BeamSearch)
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:579]: Same suggestion here ⬇️

```suggestion
BeamSearch.register("stochastic", constructor="stochastic")(BeamSearch)
```

File: allennlp/nn/samplers/sampler.py [hunk 16]
```diff
@@ -9,6 +9,10 @@ class Sampler(Registrable):
     """
 
     def __call__(
-        self, logits: torch.Tensor, num_samples: int = 1, with_replacement: bool = True
+        self,
+        log_probs: torch.Tensor,
+        perturbed_log_probs: torch.Tensor = None,
+        num_samples: int = 1,
+        with_replacement: bool = True,
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/nn/samplers/sampler.py:16]: Would need to be changed for all the subclasses too.

[the remark this answers is not in the document]
Comment (epwalsh) [reply allennlp/nn/samplers/sampler.py:16]: But sampling with replacement doesn't even make sense for Gumbel, right? So maybe we should just remove this argument entirely. It could be a parameter in the `__init__` method for samplers that it makes sense for.

**Comment by epwalsh** [comment]
In `samplers/__init__.py` we are missing Gumbel.

File: allennlp/nn/samplers/sampler.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jvstokes) [reply allennlp/nn/samplers/sampler.py:16]: You can change the with/without replacement for the other samplers, but stochastic is always without replacement. I'll switch it to be in the Init method.

File: allennlp/nn/samplers/samplers.py [hunk 17]
```diff
@@ -109,17 +135,80 @@ def __call__(
...
+    `num_samples` is the number of instances to sample
+
+    Returns a tensor of the top `num_samples` perturbed log probabilities and their
+    indices within `log_probs`.
+
+    Registered as a `Sampler` with name "gumbel-max".
+    """
+
+    def __init__(self):
+        self.num_samples = 1
```
Comment (epwalsh) [inline allennlp/nn/samplers/samplers.py:175]: I don't think we need this.
```suggestion
```

**Comment by epwalsh** [comment]
We should have a check here that `last_true_log_probabilities` is also finite: https://github.com/allenai/allennlp/pull/4724/files#diff-ccf1daaf744838c9a6be618bb929927a249b6c6dfeb6dc37fada88e660a3c5a1R451

**Comment by epwalsh** [comment]
Closed via #4768
